### PR TITLE
feature(scylla-bench refactor): refactor usage of scylla-bench

### PR DESCRIPTION
**Pull request by juliayakovlev**
Move the implementation of scylla-bench usage to it's own file.
Read the output and supply stats for Prometheus.

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

### CODE REVIEW

File: sdcm/loader.py [hunk 1]
```diff
@@ -77,3 +108,81 @@ def run(self):
...
+    def skip_line(line: str) -> bool:
+        # If line starts with 'total,' - skip this line
+        return True if not 'total,' in line else False
+
+    @staticmethod
+    def split_line(line: str) -> list:
+        return [element.strip() for element in line.split(',')]
+
+
+class ScyllaBenchStressExported(StressExporter):
```
Comment (fruch) [inline sdcm/loader.py:143]:
```suggestion
class ScyllaBenchStressExporter(StressExporter):
```

**Comment by fruch** [comment]
How about doing the extra mile and putting it under docker while changing this ? Or this is plan in next steps ?

**Comment by juliayakovlev** [comment]
> How about doing the extra mile and putting it under docker while changing this ? Or this is plan in next steps ?

There is separate task for docker< I think. It's not the part of this task

File: sdcm/loader.py (hunk 1 above)
Comment (juliayakovlev) [reply sdcm/loader.py]: thanks. fixed

File: sdcm/utils/common.py [hunk 2]
```diff
@@ -2058,3 +2058,19 @@ def parse_nodetool_listsnapshots(listsnapshots_output: str) -> defaultdict:
...
     return snapshots_content
+
+
+def convert_metric_to_ms(metric: str) -> float:
+    # Convert metric value to ms and float
+
+    if metric.endswith('µs'):
+        metric_converted = float(metric.rstrip('µs')) / 1000
+    elif metric.endswith('ms'):
+        metric_converted = float(metric.rstrip('ms'))
```
Comment (enaydanov) [inline sdcm/utils/common.py:2069]: Can it raise `ValueError` too?

**Review by enaydanov — CHANGES_REQUESTED** [review]

File: sdcm/scylla_bench_thread.py [hunk 3]
```diff
@@ -0,0 +1,265 @@
...
+# the Free Software Foundation; either version 3 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+#
+# See LICENSE for more details.
+#
+# Copyright (c) 2019 ScyllaDB
```
Comment (bentsi) [inline sdcm/scylla_bench_thread.py:12]: fix preamble

File: sdcm/sct_events.py [hunk 4]
```diff
@@ -864,8 +868,8 @@ def __init__(self,  # pylint: disable=too-many-arguments
 
     def __str__(self):
         return f"{super().__str__()}: alert_name={self.alert_name} type={self.type} start={self.start} "\
-               f"end={self.end} description={self.description} updated={self.updated} state={self.state} "\
-               f"fingerprint={self.fingerprint} labels={self.labels}"
+            f"end={self.end} description={self.description} updated={self.updated} state={self.state} "\
```
Comment (bentsi) [inline sdcm/sct_events.py:871]: same here

File: sdcm/loader.py [hunk 5]
```diff
@@ -12,62 +12,93 @@
...
+        self.stress_operation = stress_operation
+        self.stress_log_filename = stress_log_filename
+        gauge_name = self.create_metrix_gauge()
+        self.stress_metric = self.METRICS_GAUGES[gauge_name]
         self.instance_name = instance_name
         self.loader_idx = loader_idx
         self.cpu_idx = cpu_idx
+        self.metrics_positions = self.merics_position_in_log()
+
+    @abc.abstractmethod
```
Comment (bentsi) [inline sdcm/loader.py:52]: better "@abstractmethod"

File: sdcm/loader.py [hunk 6]
```diff
@@ -12,62 +12,93 @@
...
+    lat_perc_999: int
+    lat_max: int
+    errors: int
 
-        super(CassandraStressExporter, self).__init__()
+
+class StressExporter(FileFollowerThread, metaclass=abc.ABCMeta):
+    METRICS_GAUGES = {}
+
+    def __init__(self, instance_name, metrics, stress_operation, stress_log_filename, loader_idx, cpu_idx=1):  # pylint: disable=too-many-arguments
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/loader.py]: done

File: sdcm/loader.py [hunk 7]
```diff
@@ -12,62 +12,93 @@
...
-    def set_metric(self, name, value):
-        self.cs_metric.labels(0, self.instance_name, self.loader_idx, self.cpu_idx, name).set(value)
+    @abc.abstractmethod
+    def create_metrix_gauge(self) -> str:
+        ...
 
-    def clear_metrics(self):
-        if self.cs_metric:
-            for metric_name in ['ops', 'lat_mean', 'lat_med', 'lat_perc_95', 'lat_perc_99', 'lat_perc_999', 'lat_max', 'errors']:
+    def set_metric(self, name, value) -> None:
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/loader.py]: done

File: sdcm/loader.py [hunk 8]
```diff
@@ -12,62 +12,93 @@
...
-                lat_max = float(cols[10])
-                errors = int(cols[13])
+                ops = float(cols[self.metrics_positions.ops])
+                lat_mean = convert_metric_to_ms(cols[self.metrics_positions.lat_mean])
+                lat_med = convert_metric_to_ms(cols[self.metrics_positions.lat_med])
+                lat_perc_95 = convert_metric_to_ms(cols[self.metrics_positions.lat_perc_95])
+                lat_perc_99 = convert_metric_to_ms(cols[self.metrics_positions.lat_perc_99])
+                lat_perc_999 = convert_metric_to_ms(cols[self.metrics_positions.lat_perc_999])
+                lat_max = convert_metric_to_ms(cols[self.metrics_positions.lat_max])
+                errors = int(cols[self.metrics_positions.errors])
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/loader.py]: changed

File: sdcm/loader.py [hunk 9]
```diff
@@ -77,3 +108,81 @@ def run(self):
...
+    def skip_line(self, line) -> bool:
+        # If line is not starts with numeric ended by "s" - skip this line.
+        # Example:
+        #    Client compression:  true
+        #    1.004777157s                2891    28910        0  67.829759ms      64.290815ms      58.327039ms      4.653055ms       3.244031µs       1.376255µs
+
+        if not line or line == '\n':
+            return True  # skip the line
+
+        line_splitted = line.split()
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/loader.py]: fixed

File: sdcm/loader.py [hunk 10]
```diff
@@ -77,3 +108,81 @@ def run(self):
...
+        if not line_splitted[0].endswith('s'):
+            return True  # skip the line
+
+        try:
+            _ = float(line_splitted[0].rstrip('s'))
+            return False  # The line hold the metrics, don't skip the line
+        except:
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/loader.py]: changed

File: sdcm/loader.py [hunk 11]
```diff
@@ -77,3 +108,81 @@ def run(self):
...
+        return gauge_name
+
+    def merics_position_in_log(self) -> MetricsPosition:
+        return MetricsPosition(ops=2, lat_mean=5, lat_med=6, lat_perc_95=7, lat_perc_99=8, lat_perc_999=9,
+                               lat_max=10, errors=13)
+
+    @staticmethod
+    def skip_line(line: str) -> bool:
+        # If line starts with 'total,' - skip this line
+        return True if not 'total,' in line else False
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/loader.py]: fixed

File: sdcm/loader.py [hunk 12]
```diff
@@ -77,3 +108,81 @@ def run(self):
...
+
+        if not line or line == '\n':
+            return True  # skip the line
+
+        line_splitted = line.split()
+        if not line_splitted[0].endswith('s'):
+            return True  # skip the line
+
+        try:
+            _ = float(line_splitted[0].rstrip('s'))
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/loader.py]: changed

File: sdcm/sct_events.py [hunk 13]
```diff
@@ -828,7 +832,7 @@ def add_info(self, node, line: str, line_number: int) -> bool:
 
     def __str__(self):
         return f"{SctEvent.__str__(self)}: type={self.type} line_number={self.line_number} node={self.node}\n" \
-               f"{self.line}"
+            f"{self.line}"
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/sct_events.py]: done

File: sdcm/scylla_bench_thread.py [hunk 14]
```diff
@@ -0,0 +1,265 @@
...
+
+from __future__ import print_function
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/scylla_bench_thread.py]: removed

File: sdcm/utils/common.py [hunk 15]
```diff
@@ -2058,3 +2058,19 @@ def parse_nodetool_listsnapshots(listsnapshots_output: str) -> defaultdict:
             line_splitted = line.split()
             snapshots_content[line_splitted[0]].append(SnapshotDetails(line_splitted[1], line_splitted[2]))
     return snapshots_content
+
+
+def convert_metric_to_ms(metric: str) -> float:
+    # Convert metric value to ms and float
+
+    if metric.endswith('µs'):
+        metric_converted = float(metric.rstrip('µs')) / 1000
```
[the remark this answers is not in the document]
Comment (juliayakovlev) [reply sdcm/utils/common.py]: fixed

File: sdcm/utils/common.py (hunk 2 above)
Comment (juliayakovlev) [reply sdcm/utils/common.py]: done

File: sdcm/scylla_bench_thread.py (hunk 3 above)
Comment (juliayakovlev) [reply sdcm/scylla_bench_thread.py]: fixed

File: sdcm/sct_events.py (hunk 4 above)
Comment (juliayakovlev) [reply sdcm/sct_events.py]: done

File: sdcm/loader.py (hunk 5 above)
Comment (juliayakovlev) [reply sdcm/loader.py]: fixed

File: sdcm/tester.py [hunk 16]
```diff
@@ -1195,7 +1194,7 @@ def get_stress_results(self, queue, store_results=True):
         return results
 
     def get_stress_results_bench(self, queue):
-        results = self.loaders.get_stress_results_bench(queue)
+        results = queue.get_stress_results_bench()
```
Comment (bentsi) [inline sdcm/tester.py:1197]: ??? looks strange

**Review by bentsi — APPROVED** [review]

File: sdcm/tester.py (hunk 16 above)
Comment (bentsi) [reply sdcm/tester.py:1197]: the name is misleading, but as per our conversation, it is used all over the code. So let's keep it as is...

**Pull request merged**